File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # dummy = ListNode()

        # carry = 0

        # curr = dummy

        # while l1 or l2 or carry != 0:
        #     val1 = l1.val if l1 else 0
        #     val2 = l2.val if l2 else 0
        #     currSum = val1 + val2 + carry

        #     currVal = currSum % 10
        #     carry = currSum // 10

        #     if l1:
        #         l1.val = currVal
        #         curr.next = l1
        #     elif l2:
        #         l2.val = currVal
        #         curr.next = l2
        #     else:
        #         curr.next = ListNode(currVal)

        #     curr = curr.next

        #     if l1:
        #         l1 = l1.next
            
        #     if l2:
        #         l2 = l2.next

        # return dummy.next
            

        dummy = ListNode(-1)
        curr = dummy

        carry = 0

        while l1 or l2:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            total = val1 + val2 + carry
            val = total % 10
            carry = total // 10

            if l1:
                l1 = l1.next

            if l2:
                l2 = l2.next

            curr.next = ListNode(val)
            curr = curr.next 

        if carry:
            curr.next = ListNode(carry)

        return dummy.next





            

File: 0002-add-two-numbers/0002-add-two-numbers.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var l1 = l1
        var l2 = l2
        let dummy = ListNode(-1)
        var curr = dummy
        var carry = 0

        while l1 != nil || l2 != nil || carry != 0 {
            let val1 = l1?.val ?? 0
            let val2 = l2?.val ?? 0
            let currSum = val1 + val2 + carry
            carry = currSum / 10
            let currVal = currSum % 10

            curr.next = ListNode(currVal)
            curr = curr.next!
 
            l1 = l1?.next
        
            l2 = l2?.next
        }

        return dummy.next
    }
}